lib/blamer/message.rb

```diff
@@ -0,0 +1,200 @@
+
+  class Message < Base
+    def find_or_create_bug!
+      
+      criteria = bug_search_criteria
+      bug      = if deploy
+                   # for distributed projects, we need to find either
+                   # * a bug associated with the exact deploy specified in the build attribute
+                   # * or, if the exact deploy specified does not have such a bug:
+                   #   * find any open (not fixed) bug matching the criteria;
+                   #   * if such a bug exists, use that bug and alter its deploy_id to
+                   #     the given deploy;
+                   #   * otherwise, a new bug is created.
+                   b        = Bug.transaction do
+                     environment.bugs.where(criteria.merge(deploy_id: deploy.id)).first ||
+                         environment.bugs.where(criteria.merge(fixed: false)).
+                             find_or_create!(bug_attributes)
+                   end
+                   b.deploy = deploy
+                   b
+                 else
+                   # for hosted projects, we search for any bug matching the criteria under
+                   # our current environment
+                   occurrence.message = occurrence.message.partition(/[^"]+/)[1]
+                   occurrences = environment.occurrences.where("\"occurrences\".\"metadata\" LIKE ?", "%#{occurrence.message}%").joins(:bug).where(bugs: bug_search_criteria)
+                   
+                   if occurrences.empty?
+                     environment.bugs.create! bug_attributes.merge(bug_search_criteria).merge(class_name: occurrence.bug.class_name)
+                   else
+                     occurrences.first.bug
+                   end
+                 end
+
+      # If this code works as it should, there should only be one open record of
+      # a Bug at any time. A bug occurs, a new Bug is created. A new release is
+      # out, but it doesn't fix the bug, and that same Bug is reused. A new
+      # release is out that DOES fix the bug, and the bug is closed. The bug
+      # recurs, and a new open Bug is created.
+
+      # Lastly, we need to resolve the bug if it's a duplicate.
+      bug      = bug.duplicate_of(true) if bug.duplicate?
+
+      return bug
+    end
+
+    def self.resolve_revision(project, revision)
+      commit   = project.repo.object(revision)
+      if commit.nil?
+        project.repo(&:fetch)
+        commit = project.repo.object(revision)
+      end
+      raise "Unknown revision" unless commit
+      commit.sha
+    end
+
+    protected
+
+    def bug_search_criteria
+      commit = occurrence.commit || deploy.try!(:commit)
+      raise "Need a resolvable commit" unless commit
+
+      file, line, commit = blamed_revision(commit)
+      {
+          class_name:      occurrence.bug.class_name,
+          file:             file,
+          line:             line
+          # blamed_revision:  commit.try!(:sha)
```

Comment:
We're trying to match the de-dup behaviour of gmail, which knows nothing of git. A same bug is the same bug no matter which code revision.